### PR TITLE
8307488: Incorrect weight of the first ObjectAllocationSample JFR event

### PR DESCRIPTION
**Pull request by mgronlun**
Greetings,

this PR fixes the problem with incorrect weights attribution for the initial ObjectAllocationSample event.

Testing: jdk_jfr

Thanks
Markus

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8307488](https://bugs.openjdk.org/browse/JDK-8307488): Incorrect weight of the first ObjectAllocationSample JFR event

### Reviewers
 * [Erik Gahlin](https://openjdk.org/census#egahlin) (@egahlin - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14158/head:pull/14158` \
`$ git checkout pull/14158`

Update a local copy of the PR: \
`$ git checkout pull/14158` \
`$ git pull https://git.openjdk.org/jdk.git pull/14158/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14158`

View PR using the GUI difftool: \
`$ git pr show -t 14158`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14158.diff">https://git.openjdk.org/jdk/pull/14158.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14158#issuecomment-1563286469)

### CODE REVIEW

**Review by egahlin — APPROVED** [review]

File: src/hotspot/share/jfr/recorder/service/jfrRecorderService.cpp [hunk 1]
```diff
@@ -121,6 +122,30 @@ const Thread* JfrRotationLock::_owner_thread = nullptr;
...
+    coas.do_thread(it.next());
+  }
+}
+
+static void clear_object_allocation_sampling() {
+  ClearObjectAllocationSampling coas;
+  JfrJavaThreadIterator jit;
+  iterate(jit, coas);
+  JfrNonJavaThreadIterator njit;
+  iterate(njit, coas);
```
Comment (shipilev) [inline src/hotspot/share/jfr/recorder/service/jfrRecorderService.cpp:146]: Apologies for the post-integration review, but I have a question/suggestion. This code retains the JavaThread iterator while NonJavaThreadIterator is running, which feels unnecessary. Also, if we have thousands of threads, walking these threads and doing the unconditional `clear_last_allocated_bytes` there seems redundant and heavy-weight if the relevant JFR event is off.

Comment (mgronlun) [reply src/hotspot/share/jfr/recorder/service/jfrRecorderService.cpp:146]: At this point, no recording is running, and no event settings are applied. No state exists to know if the event was enabled when the previous last recording stopped.

Comment (shipilev) [reply src/hotspot/share/jfr/recorder/service/jfrRecorderService.cpp:146]: OK, that's unfortunate. 

I would still suggest to split the scopes for Java and non-Java iterators. And maybe check if `last_allocated_bytes` is not zero (which is a rare case, especially when event is disabled) before clearing, thus replacing the unconditional write with conditional read-write.

**Pull request closed without merging**